## 1. Import library dan load dataset Iris

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

## 2. Konversi ke DataFrame, tentukan X dan y
Dataset Iris memiliki 4 fitur numerik kontinu:
- Sepal Length (panjang kelopak)
- Sepal Width (lebar kelopak)
- Petal Length (panjang mahkota)
- Petal Width (lebar mahkota)

dan 3 kelas target (`Species`): *setosa, versicolor, virginica*.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]

# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

## 3. Encode label kelas menjadi angka
`LabelEncoder` mengubah label teks menjadi angka: **0 = setosa, 1 = versicolor, 2 = virginica**.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

## 4. Split data latih dan data uji
Data dibagi **80% latih : 20% uji**. `random_state=0` membuat hasil split selalu sama tiap dijalankan.

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

## 5. Latih model Gaussian Naïve Bayes
Karena semua fitur berupa data kontinu, digunakan **GaussianNB**.

In [6]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model.fit(X_train, y_train)

## 6. Prediksi data uji dan hitung akurasi

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Akurasi Gaussian Naïve Bayes pada data uji Iris adalah **96,67%** — sama dengan akurasi KNN (K=3) pada sesi sebelumnya, karena kelas-kelas Iris memang terpisah cukup baik terutama pada fitur petal.

## 7. Prediksi data baru (list biasa) — muncul warning
Jika input diberikan sebagai **list/array biasa** (tanpa nama kolom), model tetap bisa memprediksi tetapi akan muncul **UserWarning**, karena model dilatih menggunakan DataFrame yang punya nama fitur (`iris.feature_names`), sedangkan input baru tidak punya nama fitur.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\glory\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


## 8. Prediksi data baru dengan DataFrame (tanpa warning)
Solusinya: ubah input menjadi **DataFrame** dengan nama kolom yang sama seperti data latih (`iris.feature_names`), sehingga struktur datanya konsisten dan warning hilang.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Data `[5.1, 3.5, 1.4, 0.2]` diprediksi sebagai **setosa** — sesuai dengan data pertama di dataset Iris yang memang berlabel setosa.

---
# Tugas

**1. Ikuti semua tahapan dalam modul ini** — sudah dikerjakan pada sel-sel di atas (langkah 1–8).

**2. Prediksi data baru: `[7.3, 3.8, 4.6, 2.3]` masuk ke species apa?**

In [11]:
input_data_tugas = pd.DataFrame(
    np.array([[7.3, 3.8, 4.6, 2.3]]),
    columns=iris.feature_names
)

result_tugas = NB_model.predict(input_data_tugas)
prediction_label_tugas = lab.inverse_transform(result_tugas)
print('Hasil prediksi untuk input baru adalah:', prediction_label_tugas)

Hasil prediksi untuk input baru adalah: ['virginica']


In [12]:
# Melihat probabilitas tiap kelas agar penjelasan lebih lengkap
proba_tugas = NB_model.predict_proba(input_data_tugas)
for kelas, p in zip(lab.classes_, proba_tugas[0]):
    print(f'P({kelas} | X) = {p:.6f}')

P(setosa | X) = 0.000000
P(versicolor | X) = 0.000000
P(virginica | X) = 1.000000


### Penjelasan hasil

Input `[7.3, 3.8, 4.6, 2.3]` (sepal length, sepal width, petal length, petal width) diprediksi sebagai **virginica**.

Alasannya:
- **Petal width = 2.3 cm** adalah nilai yang sangat besar; pada dataset Iris, petal width sebesar ini hampir selalu dimiliki oleh spesies **virginica** (versicolor umumnya < 1.8 cm, setosa < 0.6 cm).
- **Sepal length = 7.3 cm** juga termasuk besar, dan bunga dengan sepal sebesar ini biasanya virginica.
- Dari hasil `predict_proba`, probabilitas kelas **virginica ≈ 0.999999+**, jauh lebih tinggi dibanding versicolor (≈ 0.00000037) dan setosa (≈ 0, mendekati nol absolut), sehingga model sangat yakin (hampir 100%) bahwa data ini adalah **virginica**.
- Ini konsisten dengan cara kerja Gaussian Naïve Bayes: untuk tiap fitur, model menghitung seberapa "mungkin" nilai tersebut muncul pada distribusi normal setiap kelas (menggunakan $\mu$ dan $\sigma$ kelas tersebut), lalu mengalikan keempat probabilitas fitur dengan prior kelas. Karena nilai petal length & petal width pada input sangat cocok dengan distribusi virginica (dan sangat tidak cocok dengan setosa/versicolor), hasil akhirnya condong kuat ke virginica.

## Kesimpulan
1. Gaussian Naïve Bayes cocok untuk fitur numerik kontinu seperti pada dataset Iris, dan pada praktikum ini mencapai akurasi **96,67%**, setara dengan KNN (K=3).
2. Prediksi data baru sebaiknya dilakukan dengan mengubah input menjadi **DataFrame** bernama kolom sama seperti data latih, agar tidak muncul `UserWarning`.
3. Data baru `[7.3, 3.8, 4.6, 2.3]` diklasifikasikan sebagai **virginica** dengan keyakinan mendekati 100%, karena nilai petal length dan petal widthnya sangat besar dan khas untuk spesies tersebut.